examples/unsup-sage-rw.py

```diff
@@ -0,0 +1,152 @@
+dataset = Planetoid(path, dataset)
+data = dataset[0]
+
+num_samples = [10, 5]
+batch_size = 64
+hidden_channels = 32
+walk_length = 1
+epochs = 10
+
+subgraph_loader = NeighborSampler(
+    data.edge_index, node_idx=None,
+    sizes=[-1], batch_size=batch_size, shuffle=False,
+    num_workers=0)
+
+
+class SAGE(nn.Module):
+    def __init__(self, in_channels, hidden_channels):
+        super(SAGE, self).__init__()
+        self.num_layers = 2
+        self.convs = nn.ModuleList()
+        self.convs.append(
+            SAGEConv(in_channels, hidden_channels, normalize=True))
+        self.convs.append(
+            SAGEConv(hidden_channels, hidden_channels, normalize=True))
+
+    def forward(self, x, adjs):
+        # `train_loader` computes the k-hop neighborhood of a batch of nodes,
+        # and returns, for each layer, a bipartite graph object, holding the
+        # bipartite edges `edge_index`, the index `e_id` of the original edges,
+        # and the size/shape `size` of the bipartite graph.
+        # Target nodes are also included in the source nodes so that one can
+        # easily apply skip-connections or add self-loops.
+        for i, (edge_index, _, size) in enumerate(adjs):
+            x_target = x[:size[1]]  # Target nodes are always placed first.
+            x = self.convs[i]((x, x_target), edge_index)
+            if i != self.num_layers - 1:
+                x = F.relu(x)
+        return x
+
+    def inference(self, x_all):
+        pbar = tqdm(total=x_all.size(0) * self.num_layers)
+        pbar.set_description('Evaluating')
+        # Compute representations of nodes layer by layer, using *all*
+        # available edges. This leads to faster computation in contrast to
+        # immediately computing the final representations of each batch.
+        for i in range(self.num_layers):
+            xs = []
+            for batch_size, n_id, adj in subgraph_loader:
+                edge_index, _, size = adj.to(device)
+                x = x_all[n_id].to(device)
+                x_target = x[:size[1]]
+                x = self.convs[i]((x, x_target), edge_index)
+                if i != self.num_layers - 1:
+                    x = F.relu(x)
+                xs.append(x.cpu())
+                pbar.update(batch_size)
+            x_all = torch.cat(xs, dim=0)
+        pbar.close()
+        return x_all
+
+
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+model = SAGE(data.num_node_features, hidden_channels).to(device)
+optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
+x = data.x.to(device)
+
+
+def train(epoch):
+    model.train()
+
+    pbar = tqdm(total=data.num_nodes)
+    pbar.set_description(f'Epoch {epoch:02d}')
+
+    total_loss = 0
+
+    node_idx = torch.randperm(data.num_nodes)  # shuffle all nodes
+    train_loader = NeighborSampler(
+        data.edge_index, node_idx=node_idx,
+        sizes=num_samples, batch_size=batch_size, shuffle=False,
+        num_workers=0)
+
```

Comment:
I'm getting values in `rw_idx` that are greater than `num_nodes`. I added `data.num_nodes` to my call to `random_walk`. The values look to me to be the result of arrays reading memory they shouldn't be.

Reply:
Weird, it seems to be a pytorch issue. Could you test ```random_walk``` seperately? What is the dtype of ```data.edge_index```?

Reply:
I'll have to think about how to test `random_walk` separately. The type of `data.edge_index` is `<class 'torch.Tensor'>`.

I commented about this on a closed bug https://github.com/rusty1s/pytorch_cluster/issues/81#issuecomment-696927434. @rusty1s suggests that it has to do with self loops, but I have not had time to try it out yet.